### PR TITLE
feat: time with timezone

### LINKED ISSUE #14211: New data type to store only time with timezone
## Issue Creation Checklist

- [x] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

### Is your feature request related to a problem? Please describe.

I need data type to store only time with timezone, but I did not find the data type

### Describe the solution you'd like

The solution is to create new data type.

PostgreSQL data type to solve this problem [`here`](https://www.postgresql.org/docs/9.5/datatype-datetime.html#DATATYPE-DATETIME "PostgreSQL data types") - `time [ (p) ] with time zone`

### Why should this be in Sequelize

It's useful data type.

### Additional context

Sequelize has data type `TIME` but is store `TIME WITHOUT TIMEZONE`

### Is this feature dialect-specific?

- [ ] No. This feature is relevant to Sequelize as a whole.
- [x] Yes. This feature only applies to the following dialect(s): `postgres`

### Would you be willing to implement this feature by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [x] Yes, I have the time but I don't know how to start, I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

### PR DESCRIPTION
**Pull request by Sasha-hk**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Have you added new tests to prevent regressions?
- [x] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Closes #14211 

Create data type to store only time with timezone

### CODE REVIEW

**Comment by Sasha-hk** [comment]
I did have similar error when test the code locally. But it was with `postgres` dialect.

Can I fix this or can I just wait?

File: src/data-types.d.ts
Comment (ephys) [inline src/data-types.d.ts:341]: `zoned` should be a boolean

```suggestion
  zoned?: boolean;
```

File: src/data-types.d.ts
Comment (ephys) [inline src/data-types.d.ts:327]: This can be removed

```suggestion
```

File: test/unit/sql/data-types.test.js
Comment (ephys) [inline test/unit/sql/data-types.test.js:248]: With the current implementation, I think `DataTypes.TIME({ zoned: true })` doesn't work. Can you add a test for that?

File: src/dialects/mariadb/data-types.js
Comment (ephys) [inline src/dialects/mariadb/data-types.js:9]: This warn links to the documentation for mssql instead of mariadb

```suggestion
  const warn = BaseTypes.ABSTRACT.warn.bind(undefined, 'https://mariadb.com/kb/en/data-types/');
```

File: src/dialects/mysql/data-types.js
Comment (ephys) [inline src/dialects/mysql/data-types.js:9]: Same here, this is the proper link for MySQL:

```suggestion
  const warn = BaseTypes.ABSTRACT.warn.bind(undefined, 'https://dev.mysql.com/doc/refman/8.0/en/data-types.html');
```

File: src/dialects/mysql/data-types.js
Comment (ephys) [inline src/dialects/mysql/data-types.js:101]:
```suggestion
        warn('MySQL does not support TIME WITH TIMEZONE. Plain `TIME` will be used instead.');
```

File: src/dialects/snowflake/data-types.js
Comment (ephys) [inline src/dialects/snowflake/data-types.js:7]:
```suggestion
  const warn = BaseTypes.ABSTRACT.warn.bind(undefined, 'https://docs.snowflake.com/en/sql-reference/data-types.html');
```

**Review by ephys — CHANGES_REQUESTED** [review]
Thank you for the PR :)
A few copy-paste errors, and one potentially failing scenario, but overall looking good

Could you also add `DataTypes.TIME.ZONED` in the documentation? https://github.com/sequelize/sequelize/blob/main/docs/manual/other-topics/other-data-types.md

File: src/data-types.js
Comment (Sasha-hk) [reply src/data-types.js:472]: If I remove:

```js
this.options = options;
```

Then it cause an error:

```js
  get ZONED() {
    this._zoned = true;
    this.options.zoned = true; // <<<

    return this;
  }
```

To fix this I can add such line:

```js
  get ZONED() {
    this._zoned = true;
    this.options = {};
    this.options.zoned = true;

    return this;
  }
```

Is this a good way?

**Comment by Sasha-hk** [comment]
Okay I add documentation

File: src/data-types.js
Comment (ephys) [reply src/data-types.js:472]: Simply remove `this.options.zoned = true` too, it's unnecessary as you're already assigning `this._zoned`

File: src/dialects/db2/data-types.js
Comment (Sasha-hk) [inline src/dialects/db2/data-types.js:274]: This line is not a problem? Previously I remove this.options in constructor

File: src/dialects/mssql/data-types.js
Comment (Sasha-hk) [inline src/dialects/mssql/data-types.js:151]: The line is not a problem? Previously I remove this.options in constructor. I think it can cause an error

Comment (ephys) [reply src/dialects/mssql/data-types.js:151]: this specific line can be removed (same for the other dialects)

Comment (Sasha-hk) [reply src/dialects/mssql/data-types.js:151]: Okay

File: test/unit/sql/data-types.test.js
Comment (ephys) [inline test/unit/sql/data-types.test.js:248]:
```suggestion
      testsql('TIME({ zoned: true })', DataTypes.TIME({ zoned: true }), {
```

Comment (Sasha-hk) [reply test/unit/sql/data-types.test.js:248]: You misunderstood, i make three testsql:

- DataType.TIME
- DataType.TIME({ zoned: true })
- DataType.TIME.ZONED

I will make new commits, if third test is unnecessary I can remove it

Comment (ephys) [reply test/unit/sql/data-types.test.js:248]: I know, but two of these tests are named `TIME.ZONED`, this one should be named `TIME({ zoned: true })` as that's what is being tested (the first argument is the name of the test)

File: src/dialects/db2/data-types.js
Comment (ephys) [inline src/dialects/db2/data-types.js:279]: I hadn't noticed this line before, what is the reason for it?

File: test/unit/sql/data-types.test.js
Comment (Sasha-hk) [reply test/unit/sql/data-types.test.js]: Understand, I renamed it

Comment (ephys) [reply test/unit/sql/data-types.test.js]: perfect :)

File: src/dialects/db2/data-types.js
Comment (Sasha-hk) [reply src/dialects/db2/data-types.js:279]: In the beginning, I don't understand everything, so I took as an example what you sent first as a guide for my function request 

https://github.com/sequelize/sequelize/blob/85b0984c137dcc4c79ae5627ac7973e589b0150c/src/dialects/postgres/data-types.js#L313

Comment (Sasha-hk) [reply src/dialects/db2/data-types.js:279]: Maybe I should remove this

**Comment by ephys** [comment]
I forgot one last test: the integration one

It tests that the value you put in is properly serialized/deserialized when sent/retrieved from the database.

It goes in `test/integration/data-types.test.js`
You can take inspiration from the `TIME` one (no need to use moment):
https://github.com/sequelize/sequelize/blob/85b0984c137dcc4c79ae5627ac7973e589b0150c/test/integration/data-types.test.js#L162

File: src/dialects/db2/data-types.js
Comment (ephys) [reply src/dialects/db2/data-types.js:279]: Yes, I don't think it does anything as there doesn't seem to be a static `parse` method present in its prototype chain

**Comment by Sasha-hk** [comment]
If I need to update something, I did it. Now I go to bed, tomorrow I continue, thank you for patience :)

**Comment by ephys** [comment]
I'm not sure why test integration test is failing (to be honest it's the first time I look into that test suite). I'll assign this to myself to remind myself to take a look into it when I'm available :)

**Comment by Sasha-hk** [comment]
About when will the new version of Sequelize be released?

**Comment by ephys** [comment]
The new v6 release that includes this? Not sure, I need to find the time to figure out how to write this test

**Comment by Sasha-hk** [comment]
I'm just wondering when I will be able to use the data type. I created the data type locally so as not to stop the development of my project.

**Comment by WikiRik** [comment]
> I'm not sure why test integration test is failing (to be honest it's the first time I look into that test suite). I'll assign this to myself to remind myself to take a look into it when I'm available :)

I think we might need to rewrite the testSuccess function there. I don't think it can deal with the `{ zoned: true }`. Might be best for now to just rewrite the test so it does more in the test itself and it doesn't use the testSuccess function.

**Comment by Sasha-hk** [comment]
I tried to write the test by hand, but I don't know how else to implement it

**Comment by Sasha-hk** [comment]
I don't know if it's another error or if it caused an error before

<details>
<summary>Error:</summary>

```shell
1) [POSTGRES] DataTypes
       calls parse and stringify for TIME WITH TIMEZONE:
     invalid input syntax for type time with time zone: "{}"
  Error
      at Query.run (lib/dialects/postgres/query.js:55:25)
      at /home/petryk/dev/sequelize/lib/sequelize.js:321:28
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async PostgresQueryInterface.insert (lib/dialects/abstract/query-interface.js:297:21)
      at async model.save (lib/model.js:2442:35)
      at async Function.create (lib/model.js:1348:12)
      at async Context.<anonymous> (test/integration/data-types.test.js:199:7)
```

</details>

<details>
<summary>Test:</summary>

```js
  if (current.dialect.supports.TIME_WITH_TIME_ZONE) {
    it('calls parse and stringify for TIME WITH TIMEZONE', async () => {
      const Type = new Sequelize.TIME({ zoned: true });
      const value = '12:00:00 +02:00';
      const options = {};
      let parseHaveBeenCalled = false;
      let stringifyHaveBeenCalled = false;

      // await testSuccess(Type, moment(new Date()).format('hh:mm:ss ZZ'));
      Type.constructor.parse = value => {
        parseHaveBeenCalled = true;

        return value;
      };

      Type.constructor.prototype.stringify = () => {
        stringifyHaveBeenCalled = true;

        return Reflect.apply(Sequelize.ABSTRACT.prototype.stringify, this, arguments);
      };

      const User = current.define('user', {
        field: Type,
      }, {
        timestamps: false,
      });

      await current.sync({ force: true });

      current.refreshTypes();

      await User.create({ // <<< error here
        field: value,
      });

      await User.findAll();
      expect(parseHaveBeenCalled).to.be.equal(true);
      expect(stringifyHaveBeenCalled).to.be.equal(true);

      delete Type.constructor.parse;
      delete Type.constructor.prototype.stringify;
    });
```

</details>

**Comment by ephys** [comment]
Resolved the conflicts. If it's ok with @Sasha-hk I'll take over the branch to look into the test issue

**Comment by WikiRik** [comment]
@ephys do you want to leave this open until you start on rewriting the test suite? (after #13799 has been merged probably) or can we merge this soon?

**Comment by WikiRik** [comment]
@ephys this one is blocked by #14505 I assume?

**Comment by ephys** [comment]
Yep! That branch revamps the test suite completely

**Comment by ephys** [comment]
I've done a lot more data type research in the past months with the data type rewrite and I think we should not include support for `TIMETZ`. From my research, it is a useless type that can only bring confusion. The postgres wiki is very clear about the fact that it should never be used. 

I'm sorry it comes after all this effort, but I propose to abandon this feature

**Comment by Sasha-hk** [comment]
> I've done a lot more data type research in the past months with the data type rewrite and I think we should not include support for `TIMETZ`. From my research, it is a useless type that can only bring confusion. The postgres wiki is very clear about the fact that it should never be used.
> 
> I'm sorry it comes after all this effort, but I propose to abandon this feature

I agree with you, we can close this PR

**Pull request closed without merging**
